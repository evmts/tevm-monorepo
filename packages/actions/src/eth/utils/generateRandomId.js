/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
