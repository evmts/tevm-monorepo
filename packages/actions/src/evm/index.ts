export type {
	EvmHandler,
	EvmSetNextBlockTimestampHandler,
} from './EvmHandler.js'
export type {
	EvmJsonRpcRequest,
	EvmSetNextBlockTimestampJsonRpcRequest,
} from './EvmJsonRpcRequest.js'
export type {
	EvmJsonRpcResponse,
	EvmSetNextBlockTimestampJsonRpcResponse,
} from './EvmJsonRpcResponse.js'
export type {
	EvmParams,
	EvmSetNextBlockTimestampParams,
} from './EvmParams.js'
export type {
	EvmProcedure,
	EvmSetNextBlockTimestampProcedure,
} from './EvmProcedure.js'
export type {
	EvmResult,
	EvmSetNextBlockTimestampResult,
} from './EvmResult.js'

export { evmSetNextBlockTimestampProcedure } from './evmSetNextBlockTimestampProcedure.js'