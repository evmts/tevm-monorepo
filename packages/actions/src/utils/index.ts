export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
