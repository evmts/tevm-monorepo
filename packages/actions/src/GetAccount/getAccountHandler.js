import { createAddress } from '@tevm/address'
import { AccountNotFoundError, ForkError, InternalError } from '@tevm/errors'
import { bytesToHex } from '@tevm/utils'
import { cloneVmWithBlockTag } from '../Call/cloneVmWithBlock.js'
import { getPendingClient } from '../internal/getPendingClient.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateGetAccountParams } from './validateGetAccountParams.js'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('../GetAccount/GetAccountHandlerType.js').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('../GetAccount/TevmGetAccountError.js').TevmGetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = createAddress(params.address)
		try {
			if (params.blockTag === 'pending') {
				const mineResult = await getPendingClient(client)
				if (mineResult.errors) {
					return maybeThrowOnFail(throwOnFail, {
						errors: mineResult.errors,
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(mineResult.pendingClient, options)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			if (params.blockTag !== 'latest' && params.blockTag !== undefined) {
				const block = await vm.blockchain.getBlockByTag(params.blockTag)
				const clonedVm = await cloneVmWithBlockTag(client, block)
				if (clonedVm instanceof ForkError || clonedVm instanceof InternalError) {
					return maybeThrowOnFail(throwOnFail, {
						errors: [clonedVm],
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(
					{ ...client, getVm: () => Promise.resolve(clonedVm) },
					options,
				)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [new AccountNotFoundError(`account ${params.address} not found`)],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code = res?.codeHash !== undefined ? bytesToHex(await vm.stateManager.getContractCode(address)) : '0x'

			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
				...(params.returnStorage
					? {
							storage: Object.fromEntries(
								Object.entries(await vm.stateManager.dumpStorage(address)).map(([key, value]) => [
									`0x${key}`,
									/** @type {import('../common/Hex.js').Hex}*/ (value),
								]),
							),
						}
					: {}),
			}
		} catch (e) {
			let err = e
			// TODO this isn't clean
			// What we are doing here is checking if we threw a known error or not
			if (typeof e !== 'object' || e === null || !('_tag' in e)) {
				err = new InternalError('UnexpectedError in getAccountHandler', { cause: /** @type {any}*/ (e) })
			}
			errors.push(/** @type any*/ (err))
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
