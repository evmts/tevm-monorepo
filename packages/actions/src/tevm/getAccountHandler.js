import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAddress } from '@tevm/utils'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { validateGetAccountParams } from '@tevm/zod'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'getVm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('@tevm/errors').GetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = new EthjsAddress(hexToBytes(params.address))
		try {
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [
						createError(
							'AccountNotFoundError',
							`account ${params.address} not found`,
						),
					],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code =
				res?.codeHash !== undefined
					? bytesToHex(await vm.stateManager.getContractCode(address))
					: '0x'
			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions-types
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
			}
		} catch (e) {
			errors.push(
				createError(
					'UnexpectedError',
					typeof e === 'string'
						? e
						: e instanceof Error
						? e.message
						: 'unknown error',
				),
			)
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
