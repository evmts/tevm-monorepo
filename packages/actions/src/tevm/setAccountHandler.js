import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAccount, EthjsAddress } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { validateSetAccountParams } from '@tevm/zod'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').SetAccountHandler}
 */
export const setAccountHandler = (client, options = {}) => async (params) => {
	const { throwOnFail = options.throwOnFail ?? true } = params
	/**
	 * @type {Array<import('@tevm/errors').SetAccountError>}
	 */
	const errors = validateSetAccountParams(params)
	if (errors.length > 0) {
		return maybeThrowOnFail(throwOnFail, { errors })
	}

	const address = new EthjsAddress(hexToBytes(params.address))
	try {
		await client.vm.stateManager.putAccount(
			address,
			new EthjsAccount(
				params.nonce,
				params.balance,
				params.storageRoot && hexToBytes(params.storageRoot),
				params.deployedBytecode &&
					hexToBytes(keccak256(params.deployedBytecode)),
			),
		)
		if (params.deployedBytecode) {
			await client.vm.stateManager.putContractCode(
				address,
				hexToBytes(params.deployedBytecode),
			)
		}
		await client.vm.stateManager.checkpoint()
		await client.vm.stateManager.commit()
		// TODO offer way of setting contract storage with evm.stateManager.putContractStorage
		return {}
	} catch (e) {
		errors.push(
			createError(
				'UnexpectedError',
				typeof e === 'string'
					? e
					: e instanceof Error
					? e.message
					: 'unknown error',
			),
		)
		return maybeThrowOnFail(throwOnFail, { errors })
	}
}
