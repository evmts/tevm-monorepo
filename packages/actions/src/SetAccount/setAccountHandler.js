import { createAddress } from '@tevm/address'
import { AccountNotFoundError, InternalError } from '@tevm/errors'
import { EthjsAccount } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateSetAccountParams } from './validateSetAccountParams.js'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./SetAccountHandlerType.js').SetAccountHandler}
 */
export const setAccountHandler =
	(client, options = {}) =>
	async (params) => {
		const { throwOnFail = options.throwOnFail ?? true } = params
		/**
		 * @type {Array<import('../SetAccount/TevmSetAccountError.js').TevmSetAccountError>}
		 */
		const errors = validateSetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}

		const address = createAddress(params.address)

		/**
		 * @type {Array<Promise<any>>}
		 */
		const promises = []
		try {
			const vm = await client.getVm()

			// check if account exists
			const account = await getAccountHandler(client)({ ...params, throwOnFail: false })
			if (account.errors?.length && !(account.errors[0] instanceof AccountNotFoundError)) {
				client.logger.error('there was an unexpected error getting account', account.errors)
				throw account.errors.length > 1 ? new AggregateError(account.errors) : account.errors[0]
			}

			promises.push(
				vm.stateManager.putAccount(
					address,
					new EthjsAccount(
						params.nonce ?? account?.nonce,
						params.balance ?? account?.nonce,
						(params.storageRoot && hexToBytes(params.storageRoot)) ??
							(account?.storageRoot !== undefined && account?.storageRoot !== '0x'
								? hexToBytes(account.storageRoot)
								: undefined),
						(params.deployedBytecode && hexToBytes(keccak256(params.deployedBytecode))) ??
							(account?.deployedBytecode !== undefined ? hexToBytes(keccak256(account.deployedBytecode)) : undefined),
					),
				),
			)
			if (params.deployedBytecode) {
				promises.push(vm.stateManager.putContractCode(address, hexToBytes(params.deployedBytecode)))
			}
			// state clears state first stateDiff does not
			if (params.state) {
				await vm.stateManager.clearContractStorage(address)
			}
			const state = params.state ?? params.stateDiff
			if (state) {
				for (const [key, value] of Object.entries(state)) {
					promises.push(
						vm.stateManager.putContractStorage(
							address,
							hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (key), { size: 32 }),
							hexToBytes(value),
						),
					)
				}
			}
			const results = await Promise.allSettled(promises)
			for (const result of results) {
				if (result.status === 'rejected') {
					errors.push(new InternalError('Unable to put storage', { cause: result.reason }))
				}
			}

			if (errors.length > 0) {
				return maybeThrowOnFail(throwOnFail, { errors })
			}
			await vm.stateManager.checkpoint()
			await vm.stateManager.commit(false)

			return {}
		} catch (e) {
			errors.push(new InternalError('Unexpected error setting account', { cause: e }))
			return maybeThrowOnFail(throwOnFail, { errors })
		}
	}
