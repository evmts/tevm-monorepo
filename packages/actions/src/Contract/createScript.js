import { createAddress, createContractAddress } from '@tevm/address'
import { InternalError, InternalEvmError, InvalidBytecodeError } from '@tevm/errors'
import { prefundedAccounts } from '@tevm/node'
import { createImpersonatedTx } from '@tevm/tx'
import { getAddress, hexToBytes } from '@tevm/utils'
import { runTx } from '@tevm/vm'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'

/**
 * @internal
 * Creates a script with a randomly generated address
 * @param {import('@tevm/node').TevmNode} client
 * @param {import('@tevm/utils').Hex} [code]
 * @param {import('@tevm/utils').Hex} [deployedBytecode]
 * @param {import('@tevm/utils').Address} [to]
 * @returns {Promise<{errors?: never, address: import('@tevm/utils').Address} | {address?: never, errors: Array<Error>}>}
 */
export const createScript = async (client, code, deployedBytecode, to) => {
	const scriptAddress =
		to ??
		(() => {
			const randomBigInt = BigInt(Math.floor(Math.random() * 1_000_000_000_000_000))
			return getAddress(createContractAddress(createAddress(`0x${'6969'.repeat(10)}`), randomBigInt).toString())
		})()
	const vm = await client.getVm()

	if (deployedBytecode) {
		const setAccountRes = await setAccountHandler(client)({
			address: scriptAddress,
			deployedBytecode,
			throwOnFail: false,
		})
		if (setAccountRes.errors) {
			return {
				errors: setAccountRes.errors,
			}
		}
		return {
			address: scriptAddress,
		}
	}

	if (!code) {
		return {
			errors: [new InternalError('Cannot create script without code or deployedBytecode')],
		}
	}

	const parentBlock = await vm.blockchain.getCanonicalHeadBlock()
	const priorityFee = 0n

	const sender = createAddress(/** @type {import('@tevm/utils').Address}*/ (prefundedAccounts[0]))

	let _maxFeePerGas = parentBlock.header.calcNextBaseFee() + priorityFee
	const baseFeePerGas = parentBlock.header.baseFeePerGas ?? 0n
	if (_maxFeePerGas < baseFeePerGas) {
		_maxFeePerGas = baseFeePerGas
	}
	// TODO tons of dupe
	const dataFee = (() => {
		let out = 0n
		for (const entry of hexToBytes(code) ?? []) {
			out += vm.common.vmConfig.param('gasPrices', entry === 0 ? 'txDataZero' : 'txDataNonZero')
		}
		return out
	})()
	const baseFee = (() => {
		let out = dataFee
		const txFee = vm.common.vmConfig.param('gasPrices', 'tx')
		if (txFee) {
			out += txFee
		}
		if (vm.common.vmConfig.gteHardfork('homestead')) {
			const txCreationFee = vm.common.vmConfig.param('gasPrices', 'txCreation')
			if (txCreationFee) {
				out += txCreationFee
			}
		}
		return out
	})()
	const minimumGasLimit = baseFee + BigInt(0xffffffff)
	const gasLimitWithExecutionBuffer = (minimumGasLimit * 11n) / 10n
	try {
		const res = await runTx(vm)({
			block: parentBlock,
			tx: createImpersonatedTx({
				maxFeePerGas: _maxFeePerGas,
				maxPriorityFeePerGas: 0n,
				gasLimit: gasLimitWithExecutionBuffer,
				data: code,
				impersonatedAddress: sender,
			}),
			skipNonce: true,
			skipBalance: true,
			skipBlockGasLimitValidation: true,
			skipHardForkValidation: true,
		})
		if (res.execResult.exceptionError?.error) {
			client.logger.error('Failed to create script because deployment of script bytecode failed')
			throw new InvalidBytecodeError(res.execResult.exceptionError.error, {
				cause: /** @type {any}*/ (res.execResult.exceptionError),
			})
		}
		const deployedAddress = res.createdAddress
		if (!deployedAddress) {
			return {
				errors: [new InternalEvmError('Failed to create script')],
			}
		}
		const account = await getAccountHandler(client)({
			throwOnFail: false,
			address: /** @type {import('@tevm/utils').Address}*/ (deployedAddress.toString()),
			returnStorage: true,
		})
		if (account.errors) {
			return {
				errors: account.errors,
			}
		}
		const setAccountRes = await setAccountHandler(client)({
			...account,
			address: scriptAddress,
			throwOnFail: false,
			stateDiff: account.storage ?? {},
			deployedBytecode: account.deployedBytecode,
		})
		if (setAccountRes.errors) {
			return {
				errors: setAccountRes.errors,
			}
		}
		await vm.stateManager.deleteAccount(deployedAddress)
		return {
			address: to ?? scriptAddress,
		}
	} catch (e) {
		return {
			errors: [/** @type any*/ (e)],
		}
	}
}
