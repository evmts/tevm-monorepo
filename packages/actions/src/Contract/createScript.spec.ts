import { createAddress } from '@tevm/address'
import { Block } from '@tevm/block'
import { EvmError, EvmErrorMessage } from '@tevm/evm'
import { createTevmNode } from '@tevm/node'
import { TestERC20 } from '@tevm/test-utils'
import { PREFUNDED_ACCOUNTS, getAddress } from '@tevm/utils'
import { describe, expect, it } from 'vitest'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'
import { createScript } from './createScript.js'

describe('createScript', () => {
	const client = createTevmNode()
	const validCode = TestERC20.script({ constructorArgs: ['Name', 'Symbol'] }).code
	const invalidCode = '0x6969696969' // invalid EVM bytecode
	const validDeployedBytecode = TestERC20.deployedBytecode

	it('should create a script with valid deployedBytecode', async () => {
		const result = await createScript(client, undefined, validDeployedBytecode)
		expect(result.errors).toBeUndefined()
		expect(getAddress(result.address as any)).toBeDefined()
		const account = await getAccountHandler(client)({ address: result.address as any })
		expect(account.deployedBytecode).toBe(validDeployedBytecode)
	})

	it('should create a script with valid code', async () => {
		const result = await createScript(client, validCode)
		expect(result.errors).toBeUndefined()
		expect(getAddress(result.address as any)).toBeDefined()
		const account = await getAccountHandler(client)({ address: result.address as any })
		expect(account.deployedBytecode).toBe(validDeployedBytecode)
	})

	it('should return an error for invalid bytecode', async () => {
		const result = await createScript(client, invalidCode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should return an error when neither code nor deployed bytecode is provided', async () => {
		const result = await createScript(client)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should handle script creation failure due to EVM error', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		vm.evm.runCall = async () => {
			throw new Error('EVM error')
		}
		const result = await createScript(client, validCode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should handle setAccount failing unexpecdedly', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		vm.stateManager.putAccount = async () => {
			throw new Error('Unexpected error')
		}
		const result = await createScript(client, undefined, validDeployedBytecode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should succeed even if base fee higher than calcNextBaseFee', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		const latest = await vm.blockchain.getCanonicalHeadBlock()
		const newBlock = Block.fromBlockData(
			{
				...(latest as any),
			},
			{ common: vm.common, freeze: false },
		)
		newBlock.header.calcNextBaseFee = () => 0n
		await vm.blockchain.putBlock(newBlock)
		const result = await createScript(client, validCode)
		expect(result.errors).toBeUndefined()
		expect(getAddress(result.address as any)).toBeDefined()
		const account = await getAccountHandler(client)({ address: result.address as any })
		expect(account.deployedBytecode).toBe(validDeployedBytecode)
	})

	it('should handle vm errors', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		vm.evm.runCall = () => {
			return Promise.resolve({
				execResult: {
					returnValue: Uint8Array.from([]),
					exceptionError: new EvmError(EvmErrorMessage.REVERT),
					executionGasUsed: 420n,
				},
			})
		}
		await setAccountHandler(client)({
			address: PREFUNDED_ACCOUNTS[0].address,
			balance: 0n,
		})
		const result = await createScript(client, validCode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should handle the case where the code property did not produce any bytecode', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		const originalRunCall = vm.evm.runCall.bind(vm.evm)
		vm.evm.runCall = async (params) => {
			const res = await originalRunCall(params)
			return {
				...res,
				createdAddress: undefined as any,
			}
		}
		const result = await createScript(client, validCode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})

	it('should handle the unlikely case we are unable to get the account after successful creation', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		const originalRunCall = vm.evm.runCall.bind(vm.evm)
		vm.evm.runCall = async (params) => {
			const res = await originalRunCall(params)
			return {
				...res,
				// set createdAddress to an account that does not exist
				createdAddress: createAddress(252525252525),
			}
		}
		const result = await createScript(client, validCode)
		expect(result.errors).toBeDefined()
		expect(result.errors).toMatchSnapshot()
	})
})
