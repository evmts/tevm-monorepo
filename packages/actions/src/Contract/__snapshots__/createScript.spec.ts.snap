// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createScript > should handle script creation failure due to EVM error 1`] = `
[
  [Error: EVM error],
]
`;

exports[`createScript > should handle setAccount failing unexpecdedly 1`] = `
[
  [InternalError: Unable to put storage

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: Unexpected error
Version: 1.1.0.next-73],
]
`;

exports[`createScript > should handle the case where the code property did not produce any bytecode 1`] = `
[
  [InternalEvmError: Failed to create script

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Version: 1.1.0.next-73],
]
`;

exports[`createScript > should handle the unlikely case we are unable to get the account after successful creation 1`] = `
[
  [AccountNotFoundError: account 0x0000000000000000000000000000003AcBAd8fAD not found

Docs: https://tevm.sh/reference/tevm/errors/classes/resourcenotfounderror/
Version: 1.1.0.next-73],
]
`;

exports[`createScript > should handle vm errors 1`] = `
[
  [InvalidBytecodeError: revert

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidbytecodeerror/
Details: {"error":"revert","errorType":"EvmError"}
Version: 1.1.0.next-73],
]
`;

exports[`createScript > should return an error for invalid bytecode 1`] = `
[
  [InvalidDeployedBytecodeError: InvalidLength

Docs: https://tevm.sh/reference/tevm/errors/classes/invaliddeployedbytecodeerror/
Version: 1.1.0.next-73],
]
`;

exports[`createScript > should return an error when neither code nor deployed bytecode is provided 1`] = `
[
  [InternalError: Cannot create script without code or deployedBytecode

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Version: 1.1.0.next-73],
]
`;
